### PR TITLE
Add AnatomicalOrientation class

### PR DESCRIPTION
**Pull request by blowekamp**
This PR is a WIP, and more details will be added here when ready for discussion and review.

See #4788 for this issue and discussion of these changes.

### CODE REVIEW

**Review by dzenanz — COMMENTED** [review]
Looks good so far!

**Comment by gdevenyi** [comment]
I'm wondering if some of the wrongness in #147 can be solved through the application of this class..

**Comment by issakomi** [comment]
> I'm wondering if some of the wrongness in #147 can be solved through the application of this class..

I am afraid that nothing can solve the issue with MINC, orient filters re-order pixels, but _left_ is still _left_ etc, axis are not flipped. Even more difficult that MINC files **written** by ITK incorporate that flip, probably no way to distinguish. Maybe MINC **3** could be the solution? Maybe without dependency on HDF, if possible?

**Comment by gdevenyi** [comment]
> Maybe MINC **3** could be the solution? Maybe without dependency on HDF, if possible?

HDF5 is the reason MINC exists (to properly address some of the huge deficiencies of NIFTI).

I'll just have to finish fixing the loader to read the voxels properly then.

**Comment by blowekamp** [comment]
Please move this conversation about MINC to #147 or another place on topic.

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (blowekamp) [inline Modules/Core/Common/include/itkAnatomicalOrientation.h:94]: @dzenanz What do you think of this name "ToOrientationEnum", locally I also have a "FromOrientationEnum"? 

Making it `itk::AnatomicalOrientation::ToOrientationEnum`. You earlier suggestions to use a "typedef" (using) make since too, but the enum name seems a little redundent.

The previous was [itk::SpatialOrientation::ValidCoordinateOrientationFlags](https://itk.org/Doxygen52/html/namespaceitk_1_1SpatialOrientation.html#a8240a59ae2e7cae9e3bad5a52ea3496e).  Also the enums were of the longer form "[ITK_COORDINATE_ORIENTATION_RIP](https://itk.org/Doxygen52/html/namespaceitk_1_1SpatialOrientation.html#a8240a59ae2e7cae9e3bad5a52ea3496eaa5d3197482c4335a63a1ad99d6a9edee)" to add even more length.

Comment (dzenanz) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h:94]: Since they are inside `AnatomicalOrientation` class, we could remove `Orientation` from the name, so have `ToEnum` and `FromEnum`? The previous super-long names were quite unwieldy.

File: Modules/IO/PhilipsREC/src/itkPhilipsRECImageIO.cxx
Comment (seanm) [inline Modules/IO/PhilipsREC/src/itkPhilipsRECImageIO.cxx:32]: The idea is to rename this without DICOM, correct?

Comment (blowekamp) [reply Modules/IO/PhilipsREC/src/itkPhilipsRECImageIO.cxx:32]: Thanks. I don't think this module is enabled by default.

Comment (blowekamp) [reply Modules/IO/PhilipsREC/src/itkPhilipsRECImageIO.cxx]: Done

File: Modules/Core/Common/test/itkAnatomicalOrientationGTest.cxx
Comment (blowekamp) [inline Modules/Core/Common/test/itkAnatomicalOrientationGTest.cxx:227]: The expected result is in correctly defined here. The following is the output from the [test failure](https://open.cdash.org/tests/1681790452):

/Users/runner/work/1/s/Modules/Core/Common/test/itkAnatomicalOrientationGTest.cxx:227: Failure
Expected equality of these values:
  itk_rai.GetTerms()
    Which is: { right-to-left, anterior-to-posterior, inferior-to-superior }
  expected_terms
    Which is: { left-to-right, anterior-to-posterior, inferior-to-superior }

**Comment by blowekamp** [comment]
@dzenanz Almost there. There is currently both the ToEnum, and FromEnum. The class "AnatomicalOrientation" can be seen to hold the representation of the orientation, and provide conversion with the Enum's, direction matrix, and perhaps the preferred handling it as 3 components.

There is still the ToEnum string representation left over from the original DICOMOrientation class. That is `AnatomicalOreination("LPS")` construction and `AnatomicalOrientation(OE::LPS).ToString() == "LPS"`. Anatomicalorientation::operator<< produce something like "left-to-right, anterior-to-posterior, inferior-to-superior" now, which is unambiguous. Perhaps the string character tripplets should just be removed? However the string representation is convent with a Python interface.

**Comment by dzenanz** [comment]
> AnatomicalOreination("LPS") construction and AnatomicalOrientation(OE::LPS).ToString() == "LPS"

I would like to have this. This convenience did not exist within ITK proper before, so there is no need for some sort of backwards compatibility. I wonder whether we should have a method to convert "To" string from a "From" string and vice versa. E.g. if an application has "RAI" string as a "From" code, it can call something like `AnatomicalOrientation::ConvertToFrom("RAI")` and get "LPS". Such method would be symmetric, as the same operation would be needed to go from "From" to "To", and from "To" to "From" convention.

**Comment by blowekamp** [comment]
So your suggestion is to keep the ambiguous "ToEnum" character representation, and just add a method to convert To/From character representations?

The concern here is really when `AnatomicalOrientation("RIA")` would be used, it is not clear if the string is being interpreted as a ToEnum or a FromEnum.

**Comment by dzenanz** [comment]
If you want, you can remove the string constructor. Is there another similarly convenient way to initialize it? Something like:
```cpp
auto identityAO = AnatomicalOrientation::ToEnum("LPS");
AnatomicalOrientation ao(identityAO);
```

**Comment by dzenanz** [comment]
More generally,:
```cpp
std::string toCode = "LPS";
auto identityAO = AnatomicalOrientation::ToEnum(toCode);
AnatomicalOrientation ao(identityAO);
```

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (issakomi) [inline Modules/Core/Common/include/itkAnatomicalOrientation.h:426]: This `#ifdef ITK_FUTURE_LEGACY_REMOVE` is unreachable inside `#ifndef ITK_FUTURE_LEGACY_REMOVE` block.

`[[deprecated]]`, not `[deprecated]`

For a function the syntax is e.g. `[[deprecated]] void f();`

Just FYI, with this changes or without -- you know my opinion about all this stuff. I got a notification and looked, I will unsubscribe from the PR to avoid further notifications, no need to respond.

**Comment by dzenanz** [comment]
What is the status of this PR? I would like it to be finished/merged before it becomes stale, as PR revival usually involves extra effort.

File: Modules/Filtering/ImageGrid/include/itkOrientImageFilter.h
Comment (dzenanz) [inline Modules/Filtering/ImageGrid/include/itkOrientImageFilter.h:312]: RIP in From notation should be LSA in To notation, not LPS?

**Comment by blowekamp** [comment]
@dzenanz Shall we have the SpatialOrientation filter changes be in another PR?

**Comment by dzenanz** [comment]
I am fine with that. And that might even be better, to not hold up this PR.

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (blowekamp) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h]: Thank you . I have addressed this.

File: Modules/Filtering/ImageGrid/include/itkOrientImageFilter.h
Comment (blowekamp) [reply Modules/Filtering/ImageGrid/include/itkOrientImageFilter.h]: No longer including filter in this PR.

**Comment by blowekamp** [comment]
When Legacy Remove is enabled deprecated warnings are generated causing one of the ITK Linux builds to failed. What is the expected behavior here?

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (N-Dekker) [inline Modules/Core/Common/include/itkAnatomicalOrientation.h:64]: Are these enumerators meant for FUTURE_REMOVE?

Comment (blowekamp) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h:64]: I don't particularly care. As noted by the conditional these terms are ambiguous, and the opposite of the the current SpatialOrientationEnums::[CoordinateTerms](https://itk.org/Doxygen/html/group__ITKCommon.html#gaeebc5f41f756b247ec3e5179b03102f9). The terms below like "RightToLeft" a clear an unambiguous and should be used.

Whatever is easiest for the community with these terms is likely best.

**Review by dzenanz — APPROVED** [review]

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (N-Dekker) [inline Modules/Core/Common/include/itkAnatomicalOrientation.h:445]: "FromFrom", is that intended? It looks a bit funny 🤷

Comment (blowekamp) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h:445]: Yes. There is "ToEnum" and "FromEnum" types defined, the the string representation are called "To String Encoding" and "From String Encoding". Then "Create From X" was the prefix for these explicit constructor-like methods.

Do you have another suggestion?

Comment (N-Dekker) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h:445]: > Do you have another suggestion?

Sorry, no! I haven't entirely studied your proposal, I just wanted to be sure that "FromFrom" is not a typo! (Other users might also find it confusing, I don't know.)

By the way I find it hard to review a PR of 1000+ changed lines of code and 10+ commits. (Especially when it comes to _approving_ the entire PR.) I don't know about this one specifically, but in general I would suggest to try to keep PRs a bit smaller. (And then do follow-up PRs, if necessary.) Anyway, that's just a general comment, it may be too late for this specific PR!

Comment (blowekamp) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h:445]: Yes, this PR has gotten a big unwieldy and likely should be squashed into just a couple commits. But due to the requests, I wanted to keep the change set until it was close, which I guess it is now.

I'd recommended just reviewing the new itkAnatomicalOrientation.h file closely.

**Comment by blowekamp** [comment]
Current warning is due to testing of the deprecated methods:
```
2024-09-05T14:28:44.5388532Z /home/vsts/work/1/s/Modules/Core/Common/test/itkAnatomicalOrientationGTest.cxx:244:73: warning: 'itk::AnatomicalOrientation::AnatomicalOrientation(itk::AnatomicalOrientation::LegacyOrientationType)' is deprecated: Use the AnatomicalOrientation::FromEnum type instead. [-Wdeprecated-declarations]
2024-09-05T14:28:44.5389330Z   244 |   itk::AnatomicalOrientation itk_rai(SOE::ITK_COORDINATE_ORIENTATION_RAI);
2024-09-05T14:28:44.5389692Z       |                                                                         ^
2024-09-05T14:28:44.5390384Z In file included from /home/vsts/work/1/s/Modules/Core/Common/test/itkAnatomicalOrientationGTest.cxx:22:
2024-09-05T14:28:44.5390862Z /home/vsts/work/1/s/Modules/Core/Common/include/itkAnatomicalOrientation.h:424:3: note: declared here
2024-09-05T14:28:44.5391276Z   424 |   AnatomicalOrientation(LegacyOrientationType legacyOrientation);
```

Anyone know other locations in ITK which test deprecated methods with out warnings?

**Comment by dzenanz** [comment]
This might be the answer you are looking for: https://github.com/InsightSoftwareConsortium/ITK/blob/fb61d2cadc33bf50d1f0ec5b842b70a1a7021d0e/Modules/Core/Common/test/itkSTLThreadTest.cxx#L18-L20

**Comment by dzenanz** [comment]
Taking a look at the definition in `itkMacro.h`, it might not.

File: Modules/Core/Common/test/itkAnatomicalOrientationGTest.cxx
Comment (dzenanz) [inline Modules/Core/Common/test/itkAnatomicalOrientationGTest.cxx:22]: Shouldn't this be `#ifdef`, not `#ifndef`?

Comment (blowekamp) [reply Modules/Core/Common/test/itkAnatomicalOrientationGTest.cxx:22]: If ITK_FUTURE_LEGACY_REMOVE is true then the method with the depreciation is removed, along with the test of it.

**Review by dzenanz — APPROVED** [review]
This is primarily rebase and squash? With removal of `#ifdef ITK_AMBIGUOUS_TO_COORDINATE_ENUMS`?

**Comment by blowekamp** [comment]
> This is primarily rebase and squash? With removal of `#ifdef ITK_AMBIGUOUS_TO_COORDINATE_ENUMS`?

Yes

**Comment by dzenanz** [comment]
Mark as "ready for review"?

**Comment by blowekamp** [comment]
The Documentation still needs work.

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (thewtex) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h:437]: Reading this class and the implementation for the first time, everything is quite good, but I was also confused by the `GetAsToStringEncoding`, `GetAsFromStringEncoding`, `CreateFromToStringEncoding`, `CreateFromFromStringEncoding`. Possible improvement: `GetPositiveStringEncoding`, `GetNegativeStringEncoding`, CreateFromPositiveStringEncoding`, `CreateFromNegativeStringEncoding`.

**Review by thewtex — COMMENTED** [review]
Very nice!!

Inline comment regarding method naming.

Wish list, could come later: `*Long*` methods to work with the names as used in `PrintSelf`.

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (blowekamp) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h:437]: That seems like an improved convention, it would also follow to rename "ToEnum"->"PositiveEnum" and "FromEnum"->"NegativeEnum".

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (blowekamp) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h]: I have removed the conditional.

**Comment by blowekamp** [comment]
> Wish list, could come later: `*Long*` methods to work with the names as used in `PrintSelf`.

The class has an "operator<<" so it work sells as a ivar for ITK classes derived from LightObject with a PrintSelf methods. Could you please explain this further or be more specific.

**Review by dzenanz — APPROVED** [review]

**Comment by thewtex** [comment]
> The class has an "operator<<" so it work sells as a ivar for ITK classes derived from LightObject with a PrintSelf methods.

Yes, that is nice.

> Could you please explain this further or be more specific.

I was thinking:

```cxx
  static const std::map<PositiveEnum, std::string> &
  GetCodeToLongString();
// returns "left-to-right", for example

  static const std::map<std::string, PositiveEnum> &
  GetStringToCode()
// has "left-to-right", etc. keys
```

**Comment by dzenanz** [comment]
Brad, do you plan to get around to Matt's suggestion soon? If not, I think we should merge this PR as-is and tackle the long string codes later.

**Comment by blowekamp** [comment]
There was a complication with Matt suggestion, as I recall it may have been that should happen with the individual axes enums and not the three-pul. 

There is a todo to improve the documentation. Given the confusion around these term, it is perhaps equally importation as the code ( even it the code now appears to be unambiguous in its terminology.)

But sure, merge it.

**Comment by dzenanz** [comment]
I am not trying to pressure you into merging. I am merely concerned about this big and important PR getting stale with time.

**Comment by blowekamp** [comment]
It has a been rebased to renew the freshness, and help reduce potential conflicts.

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (dzenanz) [inline Modules/Core/Common/include/itkAnatomicalOrientation.h:37]: a enumerated -> an enumerated

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (N-Dekker) [inline Modules/Core/Common/include/itkAnatomicalOrientation.h:432]: Please consider declaring those constructors `explicit` (to avoid implicit type conversion).

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (N-Dekker) [inline Modules/Core/Common/include/itkAnatomicalOrientation.h:433]: Please use `static_cast<T>(x)` instead of `T(x)`, for built-in types and enum types.

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (N-Dekker) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h:432]: OK, so you did consider 😃

File: Modules/Core/Common/include/itkAnatomicalOrientation.h
Comment (blowekamp) [reply Modules/Core/Common/include/itkAnatomicalOrientation.h]: done

**Review by dzenanz — APPROVED** [review]

**Review by thewtex — APPROVED** [review]
:champagne:

**Comment by blowekamp** [comment]
I'm planning on merging this tomorrow unless there are further comments.

**Comment by N-Dekker** [comment]
@blowekamp Shall I still have a look before the merge?

File: Modules/Core/Common/src/itkAnatomicalOrientation.cxx
Comment (blowekamp) [reply Modules/Core/Common/src/itkAnatomicalOrientation.cxx:199]: The math or the general input and output of the method?

Comment (N-Dekker) [reply Modules/Core/Common/src/itkAnatomicalOrientation.cxx:199]: So you are creating this multi-map, in order to find the direction matrix element that has the highest absolute value, right?

Comment (blowekamp) [reply Modules/Core/Common/src/itkAnatomicalOrientation.cxx:199]: Yes, that appears to be what is going on. As the comment says it was mostly a copy of the existing working method in  itk::SpatialObjectAdaptor::FromDirectionCosines

Comment (N-Dekker) [reply Modules/Core/Common/src/itkAnatomicalOrientation.cxx:199]: OK, thanks, and the math itself? How does it work? Is it a known approach?

**Comment by dzenanz** [comment]
🎆 🎊 🥳

**Comment by N-Dekker** [comment]
@blowekamp There appear some compile errors at the CI (master branch), if you know how to fix them, please do 😃 At https://open.cdash.org/viewBuildError.php?buildid=9981860
```

/Users/builder/externalModules/IO/PhilipsREC/src/itkPhilipsRECImageIO.cxx:713:61: error: no member named 'FromEnum' in 'itk::AnatomicalOrientation'
  AnatomicalOrientation coord_orient(AnatomicalOrientation::FromEnum::RSA);
                                     ~~~~~~~~~~~~~~~~~~~~~~~^
[CTest: warning matched] /Users/builder/externalModules/IO/PhilipsREC/src/itkPhilipsRECImageIO.cxx:713:37: warning: parentheses were disambiguated as a function declaration [-Wvexing-parse]
  AnatomicalOrientation coord_orient(AnatomicalOrientation::FromEnum::RSA);
                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[CTest: warning matched] /Users/builder/externalModules/IO/PhilipsREC/src/itkPhilipsRECImageIO.cxx:713:38: note: add a pair of parentheses to declare a variable
  AnatomicalOrientation coord_orient(AnatomicalOrientation::FromEnum::RSA);
                                     ^
      
 
```

**Comment by blowekamp** [comment]
Follow up: #4899

**Comment by jhlegarreta** [comment]
Seems like there's an ITK Windows Python warning introduced by this PR:
```
D:\a\1\s-build\Wrapping\Modules\ITKCommon\itkAnatomicalOrientationPython.cpp(3971):
 warning C4996: 'itk::AnatomicalOrientation::AnatomicalOrientation':
 Use the AnatomicalOrientation::FromEnum type instead.
```

e.g.
https://dev.azure.com/itkrobotwindowpython/ITK.Windows.Python/_build/results?buildId=4113&view=logs&j=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&t=77aad734-2057-5694-9ae2-ee1f5f26eae8&l=4288

Not sure how the below block could be handled better to avoid the warning:
https://github.com/InsightSoftwareConsortium/ITK/pull/4804/files#diff-0271cfd0418afac76ae9a99eb3c59923eac04cdfc032bbf51c3bc59c88ca1c81R441-R443

**Comment by blowekamp** [comment]
> [](https://open.cdash.org/viewBuildError.php?type=1&buildid=9988068)
> 
> Not sure how the below block could be handled better to avoid the warning: https://github.com/InsightSoftwareConsortium/ITK/pull/4804/files#diff-0271cfd0418afac76ae9a99eb3c59923eac04cdfc032bbf51c3bc59c88ca1c81R441-R443

Neither am I. Is the build/wrapping defining ITK_LEGACY_SILENT when compiling?

**Comment by N-Dekker** [comment]
@blowekamp Please note that the warning still uses the term `FromEnum`. Didn't you replace that by Positive/Negative Enum?

**Comment by blowekamp** [comment]
> Please note that the warning still uses the term `FromEnum`. Didn't you replace that by Positive/Negative Enum?

Yes, good catch the text in the warning should be updated as you recommend. But the issue is that the warning should not be occurring.

**Comment by blowekamp** [comment]
Perhaps the bigger problem on the nightly build:
```
/Users/builder/external[Modules/IO/PhilipsREC/src/itkPhilipsRECImageIO.cxx:721](https://github.com/InsightSoftwareConsortium/ITK/blob/0883d0c3b7967b9319ad9ac3a3d15fae576e382c/Modules/IO/PhilipsREC/src/itkPhilipsRECImageIO.cxx#L721):20: error: no viable overloaded '='
      coord_orient = SpatialOrientationEnums::ValidCoordinateOrientations::ITK_COORDINATE_ORIENTATION_RAI;
      ~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[CTest: warning matched] /Users/builder/external[Modules/Core/Common/include/itkAnatomicalOrientation.h:53](https://github.com/InsightSoftwareConsortium/ITK/blob/0883d0c3b7967b9319ad9ac3a3d15fae576e382c/Modules/Core/Common/include/itkAnatomicalOrientation.h#L53):24: note: candidate function (the implicit copy assignment operator) not viable: no known conversion from 'itk::SpatialOrientationEnums::ValidCoordinateOrientations' to 'const itk::AnatomicalOrientation' for 1st argument
class ITKCommon_EXPORT AnatomicalOrientation
                       ^
```

@seanm may be compiling with LEGACY_REMOVE enabled so this conversion function is removed.

**Comment by dzenanz** [comment]
Most of `Ryzenator`'s and `Corista`'s builds have `ITK_LEGACY_REMOVE:BOOL=ON`.

**Comment by blowekamp** [comment]
> Most of `Ryzenator`'s and `Corista`'s builds have `ITK_LEGACY_REMOVE:BOOL=ON`.

But they don't have PhilipsREC module enabled?

**Comment by dzenanz** [comment]
> But they don't have PhilipsREC module enabled?

Ah, that is a non-default module, which I indeed did not turn on.

**Pull request merged**
